### ISSUE TITLE
Console scripts not available during build

### ISSUE THREAD

**Issue by kumaraditya303** [pypa/build#214]
When building, console scripts paths are not injected into the path hence console scripts are not available during build.

Example:
When cmake is installed, `cmake.exe` is not on path, leading to error during when when `cmake` is invoked with subprocess.

**Comment by henryiii** [comment]
For cmake, you can `import cmake` and it has a utility for calling CMake there. (not a general solution)

**Comment by FFY00** [comment]
`PATH` should be preserved when building in isolation. Can you provide the steps to reproduce?

**Comment by henryiii** [comment]
I literally ran into this immediately after posting! :)

I'm hitting it in https://github.com/scikit-hep/awkward-1.0/pull/683, but you can just use the cmake_example from pybind11:

```bash
docker run -it python:3.9 bash
git clone https://github.com/pybind/cmake_example
pip install build
python -m build
```

Will produce:

```
error: [Errno 2] No such file or directory: 'cmake'
```

The workaround I mentioned does work:

```python
try:
    import cmake
    CMAKE = os.path.join(cmake.CMAKE_BIN_DIR, "cmake")
except ImportError:
    CMAKE = "cmake"

.... # Replace "cmake" with CMAKE in calls below
```

But it should be in the path.

**Comment by kumaraditya303** [comment]
I gave CMake just for an example it to be easily reproducible, as far as I think during building the console scripts should be on the `PATH` @FFY00

**Comment by gaborbernat** [comment]
You're right, per PEP-517

> All command-line scripts provided by the build-required packages must be present in the build environment's PATH. 

The question is why is not in your case.

**Comment by kumaraditya303** [comment]
https://github.com/pypa/build/blob/d8a30a316e5c66cd418d7789bfca7c4aa04de687/src/build/env.py#L131 
is supressing the warning that the console scripts are not on path. @gaborbernat @FFY00

**Comment by gaborbernat** [comment]
So indeed we fail on the above requirement. 

Why?
 
- the default subprocess runner we use https://github.com/pypa/pep517/blob/master/pep517/wrappers.py#L69-L75 does not add the scripts folder to the PATH - see https://github.com/pypa/pep517/blob/master/pep517/wrappers.py#L254-L264

What can we do to fix this? 

- https://github.com/pypa/build/blob/master/src/build/env.py#L30 needs to provide a scripts_dir to the console scripts folder of the environment (note this is not ``os.path.dirname(env_executable)`` in the general case (scripts_dir and bin_dir are separate configs); we likely want to use ``sysconfig`` module to load the scripts_dir
- We can override the default runner here https://github.com/pypa/build/blob/master/src/build/__init__.py#L162, to prepend scripts_dir to the PATH env-var.

**Comment by layday** [comment]
That sounds like the right approach. For Python 3, I thought you might be able to pull the `scripts` folder from the `EnvBuilder` but `EnvBuilder.create` returns `None`; you'd have to call `EnvBuilder.ensure_directories` again which performs I/O.

**Comment by gaborbernat** [comment]
You can avoid the I/O by using sysconfig instead 👍🏻 so that's a better path.

**Comment by gaborbernat** [comment]
@FFY00 can you confirm please before I start writing code 😄

**Comment by FFY00** [comment]
That seems fine :+1:

**Comment by gaborbernat** [comment]
@FFY00 and @layday so turn out our approach to borrow pip from the global namespace will not work. When pip generates a console script it always uses the current interpreter as the she-bang line, and this is non configurable. This ends up generating scripts with a bad shebang line (on UNIX a script within ``/var/folders/jh/kx1127wj32l8c3s97131rhmr0000gp/T/build-env-cv4mltrb/bin/demo-pkg-inline`` would look like):
```
#!/Users/bgabor8/git/github/python-build/.tox/4/dev/bin/python
# -*- coding: utf-8 -*-
import re
import sys
from demo_pkg_inline import do
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(do())
```
Which then fails at startup because obviously, the pip host package does not have the required dependencies. The problem:

- the pip code part that does this https://github.com/pypa/pip/blob/master/src/pip/_internal/operations/install/wheel.py#L708-L722
- a distlib issue detailing how one can set a custom interpreter - https://bitbucket.org/pypa/distlib/issues/134/allow-specifying-the-version-information

However, because pip does not expose this interface, we now must invoke the pip install with pip being installed in the same virtual environment like the one we're targeting. We can use the current behavior for non-isolated builds though. This sadly will make isolated builds a lot slower, but needed for correct functionality. 🤷🏻 any better ideas? (making a PR to expose this configuration in pip seems out of the question because we could only use it in new pip versions)

**Comment by henryiii** [comment]
I feel like this is related to the problem with cross-compiling, where these get the wrong interpreter burned in.

**Comment by FFY00** [comment]
That is fine. It feels like the only option. We can't just fix the console scripts as that is not always implemented by simply putting the intepreter in the shebang.

I would still make a PR to pip to add support for this, and keep the current behavior in newer pip versions.

**Comment by gaborbernat** [comment]
Their most likely are going to refuse to do so per latest feedback 🥺 they want to decrease their API surface, not increase it. So I'd not be to optimistic for a such change to be accepted anytime soon 👍and that would put us in the not so good place that we have different behavior based on pip version which can be again bad 🤔 because we double our ways to mess it up.

**Comment by henryiii** [comment]
Also, the current behavior is wrong, per PEP 517, and likely will start hitting a lot of users if this starts getting exposed via cibuildwheel or added to manylinux, because things like calling "cmake" are really common in setup.py's that build things. Pip has stopped supporting Python 2.7 and 3.5, so there then might not be a version of build that works correctly on all manylinux supported Pythons.

The "interpreter burnin" only happens on scripts pip _makes_, like console entry points? There might be a way to handle those, you can collect them with importlib.metadata. Not as familiar with "scripts". 

Having a way to control the burned in interpreter will be important for cross compiling, as well, so it's something to consider as a basis for future capabilities.

**Comment by gaborbernat** [comment]
> I would still make a PR to pip to add support for this and keep the current behavior in newer pip versions.

Raised the question to pip to clarify if what we're doing here is supported - https://github.com/pypa/pip/issues/9553 (or we're abusing a functionality for something else than was designed for).

**Comment by gaborbernat** [comment]
It seems my hunch was correct. pip has no remote install functionality. One would need to design the first such interface and then implement it before considering restoring the current functionality.

**Comment by uranusjr** [comment]
Would it be easier if the script generation process is entirely divorced of pip? `build` can use pip to install, and just re-generate the scripts later. pradyunsg/installer already has an implementation, it may be worthwhile to either extract the part into a separate package or have `build` depend on `installer`.

**Comment by gaborbernat** [comment]
It's an express desire of @FFY00 to not take on any more dependencies. ``installer`` is not relevant here because only handle wheels. What we would need is ``distlib`` (see my linked issue), which could regenerate the scripts. But inspecting the environment for console scripts and fixing things seems brittle and more trouble than worth it.

**Comment by FFY00** [comment]
*sigh*

I still think you fail to understand the whole point about my minimal dependency rule. It's not just me not wanting dependencies because whatever, adding dependencies on a project that is required to bootstrap the whole environment makes it significantly harder for distro packagers, people caring about reproducibility, and people in similar positions, to bootstrap a Python environment.
This is, and will continue to be, the main focus of this project. None of this makes the project unusable for the general user, just maybe we won't be able to provide the best UX. I have said many times before that if this is a limitation people are not happy with, we could have a separate package for normal users, which could pull all the dependencies it wants.

With that said, this minimal dependency rule naturally has low impact when it comes to `installer`, as that would be a required tool to bootstrap a Python environment from scratch for most people, so it probably would be available anyway.

> `installer` is not relevant here because only handle wheels

Console scripts are independent of the source, unless they are from editable installs, which is not the case here. `installer` has the ability to generate them, it would be possible to use it here.

> Also, distlib packages some binaries (for windows) we don't want to vendor IMHO.

`installer` does too, they are required to generate console script entrypoints.

> Fixing the paths is simple on UNIX

It's not, entrypoint scripts can be implemented in very different ways, they are implementation specific. In some cases, it would be as simple as fixing the shebang path, but not all entrypoint scripts are implemented by putting the python interpreter path in the shebang.
Fxing console scripts is out of the question, we need to regenerate them.

But there is also the case of non-entrypoint scripts, for which we would need the source, so regenerating scripts is probably also out of the question.

**Comment by gaborbernat** [comment]
> I still think you fail to understand the whole point about my minimal dependency rule.

My point was that this is out of the question due to this requirement (which I accept).

> Fxing console scripts is out of the question, we need to regenerate them.
> 
> But there is also the case of non-entrypoint scripts, for which we would need the source, so regenerating scripts is probably also out of the question.

Yeah I agree on this too.

**Comment by kumaraditya303** [comment]
@FFY00 can a version with this fix be released?

**Comment by FFY00** [comment]
Just released 0.2.0.
